WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,645 @@
+/* overlay group offset in bytes */
+#define M_COMRV_GET_GROUP_OFFSET_IN_BYTES(token)   ((pOverlayOffsetTable[token.fields.overlayGroupID]) << 9)
 
 /**
 * types
 */
+typedef struct comrvStackFrame
+{
+  /* holds return address (caller is non overlay function) or return
+     offset (caller is overlay function) */
+  u32_t callerReturnAddress;
+  /* holds callee address (callee is non overlay function) or callee
+     token (callee is an overlay function) */
+  u32_t calleeToken;
+  /* holds the offset in bytes to the previous stack frame */
+  s16_t offsetPrevFrame;
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+  /* if the calleetoken is a multi group token, this field holds the
+     actual loaded token entry in the multi group table */
+  u16_t calleeMultiGroupTableEntry;
+#endif /* d_comrv_multi_group_support */
+} comrvStackFrame_t;
+
+/* overlay token fields */
+typedef struct comrvToken
+{
+  /* overlay token indication 0: address; 1: overlay token */
+  u32_t overlayAddressToken:1;
+  /* overlay group ID the function resides in */
+  u32_t overlayGroupID:16;
+  /* data/function offset within the overlay group – 4 bytes granularity */
+  u32_t offset:10;
+  /* caller thunk indication */
+  u32_t thunk:1;
+  /* reserved */
+  u32_t reserved:1;
+  /* specify the heap ID this overlay group belongs to */
+  u32_t heapID:2;
+  /* multi group indication */
+  u32_t multiGroup:1;
+} comrvTokenFields_t;
+
+/* overlay token */
+typedef union comrvOverlayToken
+{
+  u32_t              value;
+  comrvTokenFields_t fields;
+} comrvOverlayToken_t;
+
+/* overlay token */
+typedef struct comrvPropertiesFields
+{
+  u08_t locked:1;
+  u08_t data:1;
+  u08_t size:3;
+  u08_t reserved:3;
+} comrvPropertiesFields_t;
+
+typedef union comrvEntryProperties
+{
+  u08_t                   value;
+  comrvPropertiesFields_t fields;
+} comrvEntryProperties_t;
+
+/* overlay token entry */
+typedef struct comrvOverlayHeapEntry
+{
+#ifdef D_COMRV_EVICTION_LRU
+  u08_t prevIndex;
+  u08_t nextIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+  comrvEntryProperties_t properties;
+  comrvOverlayToken_t    token;
+  void*                  pAddress;
+} comrvHeapEntry_t;
+
+/* com-rv control block */
+typedef struct comrvCB
+{
+#ifdef D_COMRV_EVICTION_LRU
+  u08_t            lruIndex;
+  u08_t            mruIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+  u08_t            numOfOverlayEntries;
+  comrvHeapEntry_t overlayHeap[];
+} comrvCB_t;
 
 /**
 * local prototypes
 */
+static void  comrvUpdateHeapEntryAccess      (u08_t entryIndex);
+static u08_t comrvGetEvictionCandidates      (u08_t requestedEvictionSize, u08_t* pEvictCandidatesList);
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t token);
 
 /**
 * external prototypes
 */
+extern void  comrv_entry               (void);
```

Comment:
comrv_entry: try to remove it or add a comment that this is not user "need to write" hook

Reply:
global not supported by gas
added comment in 4b2b10c